### ISSUE TITLE
Need a way to make synchronous calls from plugin platform code to Dart code.

### ISSUE THREAD

**Issue by amirh** [flutter/flutter#28310]
This came up as part of https://github.com/flutter/flutter/issues/25329:

The Android WebView API allows passing in a delegate that is invoked on the platform thread before a page is about to be loaded, the delegate needs to synchronously return true or false to determine whether to allow the navigation or to block it.

We needed to expose the same functionality in the webview_flutter plugin, ideally by delegating the decision to the Flutter application's Dart code. But this is currently impossible due to the combination of:
1. Communication over platform channels is asynchronous.
2. Messages sent from Dart to the platform channels are only delivered to the platform thread(by posting a task to the message loop).

Even if we were to block the platform thread waiting for the Dart code to make a decision, we would not be able to receive the message back from Dart(as the platform thread is blocked) and create a deadlock.

One potential approach for overriding this is to allow "platform channels" that deliver message to platform code on a separate thread. We could then:
1. Send a message to Dart from the platform thread.
2. Block the platform thread with some thread synchronization primitive.
3. Dart code will send the response to a dedicated thread.
4. The dedicated thread will release the latch that the platform thread is waiting on.

If we do this we should probably require that the "special thread" is always different from the platform thread.

For https://github.com/flutter/flutter/issues/25329 we're currently proceeding with a workaround that makes some security compromises(which means when an app sets a navigation delegate it should only load trusted content). But we should remove that workaround as soon as this issue is resolved.

cc @cbracken @jason-simmons

**Comment by yohom** [comment]
Hi @amirh, is there any progress on this issue?

**Comment by Hixie** [comment]
I don't see how the proposal would avoid deadlocks. Wouldn't a deadlock still happen if the dedicated isolate communicated with the main Dart thread which then communicated with the platform thread?

**Comment by xster** [comment]
cc @gaaclarke since this ties in to your https://docs.google.com/document/d/1bD_tiN987fWEPtw7tjXHzqZVg_g9H95IS32Cm609VZ8 explorations (sorry google internal) on pigeon.

**Comment by gaaclarke** [comment]
This is already possible with FFI.  I have examples of it here for android and iOS: https://github.com/gaaclarke/MessagingTiming

**Comment by xster** [comment]
@iskakaushik just found this bug with some more context. Also, there's some interest in the reverse direction as well. But it doesn't have to be the same technical solution. An example is ByteDance wanting to consult a native SDK API while building a frame without incurring the thread hopping expenses. But I'd treat that direction at a lower priority.

**Comment by sdas19** [comment]
The same limitation we are also facing when we are trying to share data via content provider from flutter app to a different native app. 

The scenario is, android part of the flutter app exposes a content provider. A different native android app is communicating to the content provider. Now upon receiving the request native part of the flutter app fires up a platform channel to fetch data from the flutter side. Now only the asynchronous callback mechanism is supported by the native to flutter platform channel for data sharing and the content provider needs data to be returned in a synchronous way from the `query` method. We can achieve this by putting a latch or blocking the thread, however, when we block the thread the native c++ code in the platform channel event queue does not propagate the communication anymore (to flutter part from native) as the platform channel only works in main thread.

If anyone can think of any workaround, please jump in.

**Comment by bdlukaa** [comment]
I believe this could be done, at least on the desktop platforms. Currently, we are able to call the code in the `/windows` folder by FFI. This could be supported on the framework for at Windows and Linux, that use C++.

Also, for Android, we could make use of the [Android NDK](https://developer.android.com/ndk), and call it through FFI. We can call Java/Kotlin code from NDK (using `GetStaticMethodID`).

I don't have experience with Swift (for iOS and macOS), but I believe it's also possible.

**Comment by bdlukaa** [comment]
Also, this would be able to solve https://github.com/flutter/flutter/issues/90080
